### PR TITLE
add initial jsg rust integration

### PR DESCRIPTION
**Pull request by anonrig**
Add Rust JSG bindings and comprehensive test coverage

This PR introduces Rust bindings for JSG, enabling
Rust code to interact with V8 and implement JavaScript APIs.

Key changes:
- Add jsg crate with V8 bindings (Local/Global handles, types, resources)
- Add jsg-macros for procedural macros (jsg_struct, jsg_resource, jsg_method)
- Add jsg-test harness with proper Realm initialization for testing
- Implement comprehensive test suite for JSG bindings
- Add support for wrapping/unwrapping Rust resources in V8

The Rust JSG implementation provides type-safe V8 integration and will
enable more Rust-based APIs in workerd.

Example usage:

```rust
#[test]
fn objects_can_be_wrapped_and_unwrapped() {
    let harness = crate::Harness::new();
    harness.run_in_context(|isolate| unsafe {
        let mut lock = Lock::from_isolate(isolate);
        let instance = TestStruct {
            str: "test".to_string(),
        };
        let wrapped = instance.wrap(&mut lock);
        let mut obj: v8::Local<'_, v8::Object> = wrapped.into();
        assert!(obj.has(&mut lock, "str"));
        let str_value = obj.get(&mut lock, "str");
        assert!(str_value.unwrap().is_string());
        assert!(!obj.has(&mut lock, "test"));
        let value = "value".to_local(&mut lock);
        assert!(value.is_string());
        obj.set(&mut lock, "test", value);
        assert!(obj.has(&mut lock, "test"));
    });
}
```

### CODE REVIEW

File: src/workerd/api/node/dns.h [hunk 1]
```diff
@@ -1,51 +0,0 @@
-// Copyright (c) 2017-2022 Cloudflare, Inc.
```
Comment (jasnell) [inline src/workerd/api/node/dns.h:1]: FWIW, we shouldn't delete this immediately. Instead, the rust impl should only be enabled using an autogate until it's determined to be stable enough in production.

Comment (anonrig) [reply src/workerd/api/node/dns.h:1]: Yes, this pull request is not ready to land yet

Comment (anonrig) [reply src/workerd/api/node/dns.h:1]: This pull-request is not ready to land. It's too early to say if we need an autogate or not. Probably yes, but it's too early right now.

Comment (jasnell) [reply src/workerd/api/node/dns.h:1]: Yes, I know it's still early. Just flagging this now so that it doesn't get forgotten. I'd say that if this does move forward, an autogate will absolutely be necessary when it comes to replacing any existing built-in.

File: src/rust/api/dns.rs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+                callback: Self::parse_naptr_record_callback,
+            },
+        ]
+    }
+
+    fn class_name() -> &'static str {
+        "DnsUtil"
+    }
+
+    fn js_instance<'a>(&self, lock: &mut v8::Lock) -> Option<v8::LocalValue<'a>> {
```
Comment (jasnell) [inline src/rust/api/dns.rs]: The `v8::Lock` here is misleading. We should keep the `v8::` prefix only for things that actually come from v8.

File: src/rust/jsg/ffi.c++ [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+Local local_new_number(Isolate* isolate, double value) {
+  v8::Local<v8::Number> val = v8::Number::New(isolate, value);
+  return to_ffi(kj::mv(val));
+}
+
+Local local_new_string(Isolate* isolate, ::rust::Str value) {
+  v8::Local<v8::String> val =
+      v8::String::NewFromUtf8(isolate, value.cbegin(), v8::NewStringType::kNormal, value.size())
+          .ToLocalChecked();
```
Comment (jasnell) [inline src/rust/jsg/ffi.c++]: I know this is still WIP, but... avoid using `ToLocalChecked()`, it will crash the process if there was an error. Here and elsewhere in this file.

File: src/rust/jsg/ffi.c++ [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+  auto prototype = constructor->PrototypeTemplate();
+
+  // Signatures protect our methods from being invoked with the wrong `this`.
+  auto signature = v8::Signature::New(isolate, constructor);
+
+  auto instance = constructor->InstanceTemplate();
+
+  instance->SetInternalFieldCount(workerd::jsg::Wrappable::INTERNAL_FIELD_COUNT);
+
+  auto classname = ::workerd::jsg::v8StrIntern(isolate, kj::str(descriptor.name));
```
Comment (jasnell) [inline src/rust/jsg/ffi.c++]: Nit: one of the reasons `jsg::Lock::str(...)` is preferable is that it accepts a `kj::StringPtr` that allows creating a JS string without the intermediate kj::String allocation and copy.

File: src/rust/jsg/jsg.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <workerd/jsg/jsg.h>
+#include <workerd/jsg/modules.h>
```
Comment (jasnell) [inline src/rust/jsg/jsg.h:4]: Just fyi... I haven't gotten through everything yet so you might already have this covered somewhere but want to make sure... you'll also need to support a variation here that supports the new module registry (see: `workerd/jsg/modules-new.h`)

File: src/rust/jsg/ffi.c++ (hunk 4 above)
Comment (anonrig) [reply src/rust/jsg/ffi.c++]: we will probably need a variation that supports ::rust::Str to avoid that extra copy

File: src/rust/jsg/ffi.c++ (hunk 3 above)
Comment (anonrig) [reply src/rust/jsg/ffi.c++]: Yes, this will change. I haven't got into fixin/improving this side yet. We will focus on testing this week and hopefully by the end of the week this PR would be ready to review.

File: src/workerd/api/node/dns.h (hunk 1 above)
Comment (anonrig) [reply src/workerd/api/node/dns.h:1]: I'm adding a TODO to the description and resolving this.

File: src/rust/jsg/jsg.h (hunk 5 above)
Comment (anonrig) [reply src/rust/jsg/jsg.h:4]: Probably this pull-request will not implement this. I'll do it in a follow up.

File: src/rust/jsg/ffi.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+struct ModuleRegistry;
+struct Local;
+struct Global;
+struct Realm;
+using ModuleCallback = ::rust::Fn<Local(Isolate*)>;
+using WeakCallback = ::rust::Fn<void(Isolate*, size_t)>;
+
+struct ResourceDescriptor;
+
+// Function declarations
```
Comment (jasnell) [inline src/rust/jsg/ffi.h:26]: Nit: These could use more doc comments just in general.

**Comment by jasnell** [comment]
The C++ bits LGTM. Looked over the Rust bits too and they look fine but would rather have someone who's more of a rust expert to do the sign off.

File: src/rust/api/dns.rs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+    InvalidDnsResponse(String),
+    #[error("unknown dns parser error")]
+    Unknown,
+}
+
+impl From<DnsParserError> for jsg::Error {
+    fn from(val: DnsParserError) -> Self {
+        Self::new("Error".to_owned(), val.to_string())
+    }
+}
```
Comment (guybedford) [inline src/rust/api/dns.rs:23]: Wondering if this should be derived by a macro as well for common patterns? Can always be added later I suppose.

File: src/rust/jsg/v8.rs [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+
+use crate::Lock;
+
+#[expect(clippy::missing_safety_doc)]
+#[cxx::bridge(namespace = "workerd::rust::jsg")]
+pub mod ffi {
+    #[derive(Debug)]
+    struct Local {
+        ptr: usize,
+    }
```
Comment (guybedford) [inline src/rust/jsg/v8.rs:12]: It's a shame not to have an associated lifetime here. Do you think we could integrate one?

File: src/rust/jsg/ffi.c++ [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+  // auto& typeWrapper = static_cast<TypeWrapper&>(*this);
+
+  // ResourceTypeBuilder<TypeWrapper, T, isContext> builder(
+  //     typeWrapper, isolate, constructor, instance, prototype, signature);
+
+  // if constexpr (isDetected<GetConfiguration, T>()) {
+  //   T::template registerMembers<decltype(builder), T>(builder, configuration);
+  // } else {
+  //   T::template registerMembers<decltype(builder), T>(builder);
+  // }
```
Comment (guybedford) [inline src/rust/jsg/ffi.c++:222]: Does this need a TODO or similar?

File: src/rust/jsg-test/lib.rs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    struct MultiPropertyStruct {
+        pub name: String,
+        pub age: u32,
+        pub active: String,
+    }
+
+    #[jsg_struct]
+    struct NestedStruct {
+        pub inner: String,
+    }
```
Comment (guybedford) [inline src/rust/jsg-test/lib.rs:66]: Can we not test resource call bindgen? If not, perhaps we need macro expansion output based tests here? Would help to work through arg and ret value edge cases. I really hope we can flesh out more of these basics for function bindgen before rolling out too much code coverage on this.

File: src/rust/jsg/ffi.c++ (hunk 9 above)
Comment (anonrig) [reply src/rust/jsg/ffi.c++:222]: Yes. This will be done on a separate pr that adds type generation support.

File: src/rust/jsg/v8.rs (hunk 8 above)
Comment (anonrig) [reply src/rust/jsg/v8.rs:12]: this struct is only used to pass the data between C++ and Rust. C++ either takes a reference or an rvalue reference for almost all of those values and binding takes care of lifetime and asserts it.

Comment (guybedford) [reply src/rust/jsg/v8.rs:12]: But in Rust, I can hold a Local beyond the call scope right? So it's not a compile error but a runtime error?

Comment (anonrig) [reply src/rust/jsg/v8.rs:12]: the struct is only there to distinguish a local from a global in FFI layer. simply it's a usize - the pointer to the underlying resource and yes you can hold it.

I'll think about this more and get back to you. Thank you. This is helpful.

File: src/rust/jsg-test/lib.rs (hunk 10 above)
Comment (anonrig) [reply src/rust/jsg-test/lib.rs:66]: I'll definitely add more tests and do snapshot testing on the expanded code etc. Probably not on this pull request. If that's ok with you? We landing this pull request doesn't mean we will enable it asap. It requires 2-3 more pull requests.

File: src/rust/api/dns.rs (hunk 7 above)
Comment (anonrig) [reply src/rust/api/dns.rs:23]: Hmm, yes you're right. This could definitely be put behind a macro. I didn't fully implement errors yet, so will do a follow up for this.

File: src/rust/jsg-test/lib.rs (hunk 10 above)
Comment (guybedford) [reply src/rust/jsg-test/lib.rs:66]: Sure, as long as coverage will be expanded suitably before / along with expanding more APIs.

**Review by guybedford — APPROVED** [review]

File: src/rust/jsg/v8.rs (hunk 8 above)
Comment (guybedford) [reply src/rust/jsg/v8.rs:12]: Ah, I see what you mean in that this is just a temporary wrapper struct.

File: src/workerd/api/node/node.h [hunk 11]
```diff
@@ -207,6 +215,11 @@ void registerNodeJsCompatModules(Registry& registry, auto featureFlags) {
       }
     }
   }
+
+  if (util::Autogate::isEnabled(util::AutogateKey::RUST_BACKED_NODE_DNS)) {
+    ::workerd::rust::jsg::RustModuleRegistry r(registry);
+    ::workerd::rust::api::register_nodejs_modules(r);
+  }
```
Comment (npaun) [inline src/workerd/api/node/node.h:222]: In the future, when we have more rust modules, this name might be confusing if it's called `RUST_BACKED_NODE_DNS` but it adds all the Rust modules. Unless I'm misreading it.

Comment (anonrig) [reply src/workerd/api/node/node.h:222]: Right now only Rust backed node dns is implemented. Before implementing a new module, we'll enable this flag, make sure it rolls without any issues, and later make more progress.

File: src/rust/jsg-macros/lib.rs [hunk 12]
```diff
@@ -187,17 +189,12 @@ pub fn jsg_method(_attr: TokenStream, item: TokenStream) -> TokenStream {
...
             let self_ = jsg::unwrap_resource::<Self>(&mut lock, this);
-            match self_.#fn_name(#(#arg_refs),*) {
-                Ok(result) => args.set_return_value(result.wrap(&mut lock)),
-                Err(err) => {
-                    println!("Error: {err}");
-                    todo!("{err}");
-                }
-            }
+            let result = self_.#fn_name(#(#arg_refs),*);
+            unsafe { jsg::handle_result(&mut lock, &mut args, result) };
```
Comment (guybedford) [inline src/rust/jsg-macros/lib.rs:197]: Could we put these kinds of unsafe / system calls into a special namespace that is clear that it isn't meant to be used directly in user code?

Comment (anonrig) [reply src/rust/jsg-macros/lib.rs:197]: Isn't that the whole point of using `unsafe` keyword in Rust?
